x/gamm/twap/types/keys.go

```diff
@@ -8,6 +20,76 @@ const (
 	RouterKey         = ModuleName
 
 	QuerierRoute = ModuleName
+	// Contract: Coin denoms cannot contain this character
+	KeySeparator = "|"
 )
 
-var AlteredPoolIdsPrefix = []byte{0}
+var mostRecentTWAPsNoSeparator = "recent_twap"
+var historicalTWAPTimeIndexNoSeparator = "historical_time_index"
+var historicalTWAPPoolIndexNoSeparator = "historical_pool_index"
+
+var mostRecentTWAPsPrefix = mostRecentTWAPsNoSeparator + KeySeparator
+var historicalTWAPTimeIndexPrefix = historicalTWAPTimeIndexNoSeparator + KeySeparator
+var historicalTWAPPoolIndexPrefix = historicalTWAPPoolIndexNoSeparator + KeySeparator
+
+// TODO: make utility command to automatically interlace separators
+
+func FormatMostRecentTWAPKey(poolId uint64, denom1 string, denom2 string) []byte {
+	return []byte(fmt.Sprintf("%s%d%s%s%s%s", mostRecentTWAPsPrefix, poolId, KeySeparator, denom1, KeySeparator, denom2))
+}
+
+// TODO: Replace historical management with ORM, we currently accept 2x write amplification right now.
+func FormatHistoricalTimeIndexTWAPKey(accumulatorWriteTime time.Time, poolId uint64, denom1 string, denom2 string) []byte {
+	timeS := osmoutils.FormatTimeString(accumulatorWriteTime)
+	return []byte(fmt.Sprintf("%s%s%s%d%s%s%s%s", historicalTWAPTimeIndexPrefix, timeS, KeySeparator, poolId, KeySeparator, denom1, KeySeparator, denom2))
+}
+
+func FormatHistoricalPoolIndexTWAPKey(poolId uint64, accumulatorWriteTime time.Time, denom1 string, denom2 string) []byte {
+	timeS := osmoutils.FormatTimeString(accumulatorWriteTime)
+	return []byte(fmt.Sprintf("%s%d%s%s%s%s%s%s", historicalTWAPPoolIndexPrefix, poolId, KeySeparator, timeS, KeySeparator, denom1, KeySeparator, denom2))
+}
+
```

Comment:
For my learning, can you explain why we would need to separate this formatting this way? Both use the same data, just ordered differently.

Reply:
They do not need to appear in the struct they must appear in the key.

This is because a pool can have many pairs, not just a single pair, due to the complexity of multiple asset pools.

Reply:
we need indexes by time to prune easily, and indexes by pool id to get a historical record, by time for a pool